dogma.go

```diff
@@ -262,42 +266,84 @@ type errorMessage struct {
 	Message string `json:"message"`
 }
 
-func (c *Client) do(ctx context.Context, req *http.Request, resContent interface{}) (int, error) {
-	req = req.WithContext(ctx)
-	res, err := c.client.Do(req)
-	if err != nil {
-		return UnknownHttpStatusCode, err
-	}
-	defer func() {
+func drainupAndCloseResponseBody(body io.ReadCloser) {
+	if body != nil {
 		// drain up 512 bytes and close the body to reuse connection
 		// see also:
 		// - https://github.com/google/go-github/pull/317
 		// - https://forum.golangbridge.org/t/do-i-need-to-read-the-body-before-close-it/5594/4
-		io.CopyN(ioutil.Discard, res.Body, 512)
+		io.CopyN(ioutil.Discard, body, 512)
+
+		// close body
+		body.Close()
+	}
+}
 
-		res.Body.Close()
-	}()
+func (c *Client) do(ctx context.Context, req *http.Request, resContent interface{}) (statusCode int, err error) {
+	req = req.WithContext(ctx)
+
+	// prepare metrics
+	var metricLabels []metrics.Label
+	if c.metricCollector != nil {
+		metricLabels = []metrics.Label{
+			{Name: "method", Value: req.Method},
+			{Name: "host", Value: req.URL.Host},              // included port
+			{Name: "path", Value: req.URL.EscapedPath()},     // escaped path
+			{Name: "query", Value: req.URL.Query().Encode()}, // encoded query
```

Comment:
I think there won't be large number of identical (unique) query and path in reality. 
Moreover, metric collector is not enabled by default. In case it's, then I think user want to have a deeper look (as detailed metric as possible).
HDYT, @trustin  🤔